src/Utilities/AnsiBuilder.cs

```diff
@@ -0,0 +1,330 @@
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace WheelMUD.Utilities
+{
+    /// <summary>
+    /// Mutable String class for Ansi based strings, optimized for speed and memory while retrieving the final result
+    /// as a string. Similar use to StringBuilder, but avoid a lot of allocations done by StringBuilder
+    /// </summary>
+    public class AnsiBuilder
+    {
+        /// <summary>
+        ///     Working mutable string
+        /// </summary>
+        private char[] buffer;
+
+        private int bufferPos;
+        private int charsCapacity;
+
+        /// <summary>
+        /// Temporary string used for the Replace method
+        /// </summary>
+        private List<char> replacement;
+
+        /// <summary>
+        /// Immutable string. Generated at last moment, only if needed
+        /// </summary>
+        private string stringGenerated = "";
+
+        /// <summary>
+        /// Dirty flag for AnsiBuilder containing ansi that needs parsing
+        /// </summary>
+        private bool needsParsing;
+
+        public AnsiBuilder(int initialCapacity = 32)
+        {
+            buffer = new char[charsCapacity = initialCapacity];
+        }
+
+        public int Length => ToString().Length;
+
+        /// <summary>
+        /// Return the string
+        /// </summary>
+        public override string ToString()
+        {
+            if (needsParsing) Parse();
+            stringGenerated = new string(buffer, 0, bufferPos);
+            return stringGenerated;
+        }
+
+        /// <summary>
+        /// Reset the char array
+        /// </summary>
+        public AnsiBuilder Clear()
+        {
+            bufferPos = 0;
+            needsParsing = false;
+            return this;
+        }
+
+        /// <summary>
+        /// Append a ANSI separator.
+        /// </summary>
+        /// <param name="design">repeating char of separator</param>
+        /// <param name="color">Color of the separator</param>
+        /// <param name="bold"></param>
+        /// <param name="length">how long the separator is</param>
+        /// <returns></returns>
+        public AnsiBuilder AppendSeparator(char design = '-', string color = "white", bool bold = false, int length = 60)
```

Comment:
Ah, this reminds me of another potentially useful future direction for AnsiBuilder:  If we do want it to be AnsiBuilder that performs the `<%token%>` replacements, we probably would also want it to perform the word-wrapping as it goes.  (Once the ANSI sequences that don't take up any client rendering characters are put into the output string, they're harder to compensate for with word-wrapping.)

E.G. we'd construct a new AnsiBuilder with the target word-wrapping width as the first required parameter, and when we `ab.Append("<%red%>Hi")` then we'd append the ANSI sequence for red into the buffer, followed by "Hi" but only move a renderedCursorPosition to 2 based on the length of the non-token parts.  Every Append would break the output into words to append them one at a time until the line width would be exceeded, then insert AnsiSequences.NewLine to force the word-wrapping before the word that needed it to wrap.  I think a big downside of this approach is that if all the magic only happens in AnsiBuilder then the places where we just simply write one line of output through the connection / DataFormatter process, lose out on that magic unless we find a way to share it well.

Again I think we just need to align on where our goal posts are here and whether we have the time and focus to ensure we don't drop the ball halfway there.  :)

PS - If AnsiBuilder became aware of our target client width, then the default behavior of a method like AppendSeparator (if length was not overrided from 0 or -1) could print out nicely right to the client width.

Reply:
This would certainly be the way to go. Let the AnsiBuilder handle all the string manipulation being sent to the client.

Personally I think that all the text being sent through client should be appended in a AnsiBuilder. One liners would have a little extra coding

`
var somemessage = "this is the message"
var ab = new StringBuilder(message.Length)
ab.ToString(message)
`

but this would keep all text being sent uniform, it would also make sure the line endings are always the same.

Reply:
We could even take that one step further and give AnsiBuilder a static method that does those lines of ceremony for you, like:  `var toSend = AnsiBuilder.SingleLine("<%red%>Foo<%n%>");`
I'm definitely digging this now.